app/models/message.rb

```diff
@@ -2,18 +2,48 @@ class Message < ApplicationRecord
   include WithSoftDeletion
 
   belongs_to :discussion, optional: true
-  belongs_to :assignment, foreign_key: :submission_id, primary_key: :submission_id, optional: true
+  belongs_to :assignment, optional: true
   belongs_to :approved_by, class_name: 'User', optional: true
 
   has_one :exercise, through: :assignment
 
   validates_presence_of :content, :sender
-  validates_presence_of :submission_id, :unless => :discussion_id?
+  validate :ensure_contextualized
 
   after_save :update_counters_cache!
 
   markdown_on :content
 
+  # Visible messages are those that can be publicly seen
+  # in forums. non-direct messages are never visible.
+  scope :visible, -> () do
+    where.not(deletion_motive: :self_deleted)
+      .or(where(deletion_motive: nil))
+      .where(assignment_id: nil)
+  end
+
+  def contextualization
+    direct? ? assignment : discussion
+  end
+
+  def contextualized?
+    assignment_id.present? ^ discussion_id.present?
+  end
+
+  # Whether this message is stale, that is,
+  # targets a submission that is the latest one.
```

Comment:
This caught me off guard for a second - shouldn't it be a submission that is **not** the latest one?

Reply:
You are right! I will change it

Reply:
Done